staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -206,6 +206,42 @@ func (w *watchCache) objectToVersionedRuntimeObject(obj interface{}) (runtime.Ob
 	return object, resourceVersion, nil
 }
 
+func setCachingObjects(event *watchCacheEvent, versioner storage.Versioner) {
+	// event.Object is always not nil - for object deletions it's already
+	// set to the last state of the object with RV from the time of deletion.
+	if object, err := newCachingObject(event.Object); err == nil {
+		event.Object = object
+	} else {
+		klog.Errorf("couldn't create cachingObject from: %#v", event.Object)
+	}
+
+	if event.PrevObject == nil {
+		// Don't wrap nil PrevObject for create events.
+		return
+	}
+	if event.Object != nil {
```

Comment:
if event.Object is non-nil even in delete cases, drop this block

Reply:
or switch the condition to check the event type

Reply:
if it's not ridiculously hard, see if we can add a test that ensures the PrevObject cases we wanted to be cached are actually cached

Reply:
updated this function - I think it's much cleaner now; also added a test